### PR TITLE
CDAP-16305 fix primary key updates

### PR DESCRIPTION
**Pull request by albertshau**
Fixed the algorithm to properly handle updates to primary key
fields. This requires expanding the staging table so that it
includes the previous value of the row as well as the new value.

### CODE REVIEW

File: src/test/resources/logback-test.xml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  ~ WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+  ~ License for the specific language governing permissions and limitations under
+  ~ the License.
+  -->
+
+<configuration>
+
+  <appender name="Console" class="ch.qos.logback.core.ConsoleAppender">
+    <encoder>
+      <pattern>%d{ISO8601} - %-5p [%t:%C{1}@%L] - %m%n</pattern>
```
Comment (shifuxu0301) [inline src/test/resources/logback-test.xml:23]: Just curious and take this as an education. What is the the "logback" for?

File: src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java [hunk 2]
```diff
@@ -161,6 +154,8 @@
   private final Bucket bucket;
   private final String project;
   private final RetryPolicy<Object> commitRetryPolicy;
+  private final Map<TableId, Long> latestSeenSequence;
+  private final Map<TableId, Long> latestMergedSequence;
```
Comment (shifuxu0301) [inline src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java:158]: Quick question? are these maps thread-safe?

File: src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java [hunk 3]
```diff
@@ -431,77 +451,140 @@ private void loadStagingTable(TableId stagingTableId, TableBlob blob, int attemp
...
+     * | 1234567890     | 6             | INSERT |            |              | 0  | Alice
+     * | 1234567890     | 7             | INSERT |            |              | 1  | blob
+     * | 1234567890     | 8             | UPDATE | 1          | blob         | 1  | Bob
+     *
+     * If the primary key is the 'id' field, the merge is performed by running the following query:
+     *
+     * MERGE [target table] as T
+     * USING ($DIFF_QUERY) as D
+     * ON T.id = D._before_id
+     * WHEN MATCHED AND D._op = "DELETE
```
Comment (shifuxu0301) [inline src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java]: "DELETE"

File: src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java [hunk 4]
```diff
@@ -551,4 +635,52 @@ private void runWithRetries(ContextualRunnable runnable, TableBlob blob, String
       throw e;
     }
   }
+
+  private long getLatestSequenceNum(TableId tableId) throws InterruptedException, DeltaFailureException {
+    RetryPolicy<Object> retryPolicy = new RetryPolicy<>()
+      .withMaxDuration(Duration.of(context.getMaxRetrySeconds(), ChronoUnit.SECONDS))
+      // maxDuration must be greater than the delay, otherwise Failsafe complains
+      .withBackoff(Math.min(61, context.getMaxRetrySeconds()) - 1, 120, ChronoUnit.SECONDS)
```
Comment (shifuxu0301) [inline src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java:644]: It looks like when maxRetrySecs are less than 60s, we will end up with retry just once? Should we consider retrying more times? maybe at least 2 or 3?

File: src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java [hunk 5]
```diff
@@ -551,4 +635,52 @@ private void runWithRetries(ContextualRunnable runnable, TableBlob blob, String
...
+    try {
+      return Failsafe.with(retryPolicy).get(() -> {
+        if (bigQuery.getTable(tableId) == null) {
+          return 0L;
+        }
+
+        String query = String.format("SELECT MAX(_sequence_num) FROM %s.%s", tableId.getDataset(), tableId.getTable());
+        QueryJobConfiguration jobConfig = QueryJobConfiguration.of(query);
+        JobId jobId = JobId.of(UUID.randomUUID().toString());
+        Job queryJob = bigQuery.create(JobInfo.newBuilder(jobConfig).setJobId(jobId).build());
```
Comment (shifuxu0301) [inline src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java:660]: Just curious, we are creating a big query job with a random UUID as jobId here? Wandering why it is required to have jobId here?

File: src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java [hunk 6]
```diff
@@ -551,4 +635,52 @@ private void runWithRetries(ContextualRunnable runnable, TableBlob blob, String
...
+        queryJob.waitFor();
+        TableResult result = queryJob.getQueryResults();
+        Iterator<FieldValueList> resultIter = result.iterateAll().iterator();
+        if (!resultIter.hasNext()) {
+          return 0L;
+        }
+        FieldValue val = resultIter.next().get(0);
+        if (val.getValue() == null) {
+          return 0L;
+        }
```
Comment (shifuxu0301) [inline src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java:672]: Can we add a comment here to explain what is get(0) for? I think that is for the getting the sequence number.

File: src/test/resources/logback-test.xml (hunk 1 above)
Comment (albertshau) [reply src/test/resources/logback-test.xml:23]: this configures how log messages should be printed in unit tests. Before adding this and logback-classic as a test dependency in the pom, no log messages would show up when running the test.

File: src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java (hunk 2 above)
Comment (albertshau) [reply src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java:158]: they are not, but the methods in this class that access them are synchronized. I can add a comment

File: src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java (hunk 4 above)
Comment (albertshau) [reply src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java:644]: In this case we are executing bigquery operations, which means the most likely cause of failure is some sort of misconfiguration or permissions issue, which are unlikely to be resolved without manual intervention. So if the user has set a low retry timeout, I think it's fine if we don't do many attempts.

File: src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java (hunk 5 above)
Comment (albertshau) [reply src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java:660]: This is from the example in the docs at https://cloud.google.com/bigquery/docs/quickstarts/quickstart-client-libraries#running_the_query. I haven't tried without setting the ID to see if BQ generates one for you.

File: src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java (hunk 4 above)
Comment (shifuxu0301) [reply src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java:644]: agree. makes sense for me

File: src/test/resources/logback-test.xml (hunk 1 above)
Comment (shifuxu0301) [reply src/test/resources/logback-test.xml:23]: Thanks.

File: src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java (hunk 2 above)
Comment (shifuxu0301) [reply src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java:158]: Cool.

**Review by shifuxu0301 — APPROVED** [review]
LGTM, Just need to add additional comments and add the quote before merge.

**Pull request merged**